### PR TITLE
Fix proposal navigation visibility during voting

### PR DESCRIPTION
**Pull request by mstrasinskis**
# Motivation

Fix the missing proposal navigation buttons during a vote registration.
Sns fix will be implemented in another pr.

# Changes

- Combine the ids for navigation component from available for voting proposal ids and currently in voting process.
- ProposalNavigation cares about the ids order and sorts them (because the combined list).

# Tests

- ProposalNavigation sorts the ids
- The issue itself could be better addressed by an end-to-end test to ensure that the navigation is functioning properly even during the registration process.

# Screenshots

Screen Recording 2023-06-15 at 15 42 07

### CODE REVIEW

File: frontend/src/lib/components/proposal-detail/NnsProposal.svelte [hunk 1]
```diff
@@ -13,13 +13,43 @@
...
+    type VoteRegistrationStoreData,
+  } from "$lib/stores/vote-registration.store";
+  import { OWN_CANISTER_ID_TEXT } from "$lib/constants/canister-ids.constants";
+  import type { ProposalsStore } from "$lib/stores/proposals.store";
 
   const { store } = getContext<SelectedProposalContext>(
     SELECTED_PROPOSAL_CONTEXT_KEY
   );
 
+  // The proposal that is currently in vote registration process is not included in the filtered proposals list,
```
Comment (mstrasinskis) [inline frontend/src/lib/components/proposal-detail/NnsProposal.svelte:27]: Here is the description of the issue.

**Review by lmuntaner — APPROVED** [review]
Great, thanks!

**Comment by mstrasinskis** [comment]
The last commit should fix the navigation issue with the checkbox selected. 

@dskloetd, since you discovered it could you please verify the behaviour?

https://github.com/dfinity/nns-dapp/assets/98811342/0eba2848-656f-4775-bed7-6ce46fe1d075

**Comment by mstrasinskis** [comment]
@dskloetd extra logic that combines 2 arrays solves the current issue of the temporary unavailable for navigation proposal.

1. start voting - proposal 469 (w/ all or part of the votable neurons)
2. click "Older" , goto 468
3. click "Newer", goto 470. Proposal 469 was skipped. (screenshot)

## prod version
https://github.com/dfinity/nns-dapp/assets/98811342/f7ac438c-5e80-4c4c-9a51-391467eaaf1a

With the updated version, proposals that are currently in the voting process are still available for navigation.

**Comment by dskloetd** [comment]
* Is the issue that a proposal becomes temporarily unavailable, currently existing on mainnet? If not, why was it working before and why did it break?
* Why does the proposal disappear from the store of proposals that can be voted on while it can still be voted on? It seems to me that that needs to be fixed, rather than papered over.

**Comment by mstrasinskis** [comment]
> Is the issue that a proposal becomes temporarily unavailable, currently existing on mainnet? If not, why was it working before and why did it break?

Yes, it exists on mainnet.

> Why does the proposal disappear from the store of proposals that can be voted on while it can still be voted on? It seems to me that that needs to be fixed, rather than papered over.

Probably it's an obsolete logic (to remove it from the store) to prevent user from simultaneous voting. Before the UX was different, with the loading screen during the voting process.

**Comment by dskloetd** [comment]
> Probably it's an obsolete logic (to remove it from the store) to prevent user from simultaneous voting. Before the UX was different, with the loading screen during the voting process.

If the logic is obsolete, then we should fix that instead of working around it.
It the same behavior exists on mainnet, then maybe just merge the fix for the regression now and look into the proper fix for the existing issue separately?

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts [hunk 2]
```diff
@@ -61,7 +61,7 @@ describe("ProposalNavigation", () => {
     it("should disable previous button when it's selected", async () => {
       const po = renderComponent({
         currentProposalId: 1n,
-        proposalIds: [1n, 2n],
+        proposalIds: [0n, 1n],
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts:64]: Why this change?
Now `1n` is the last proposal so previous should be enabled and next should be disabled, right?

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts [hunk 3]
```diff
@@ -72,7 +72,7 @@ describe("ProposalNavigation", () => {
     it("should disable next when it's selected", async () => {
       const po = renderComponent({
         currentProposalId: 1n,
-        proposalIds: [0n, 1n],
+        proposalIds: [1n, 2n],
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts:75]: Same

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts [hunk 4]
```diff
@@ -81,33 +81,31 @@ describe("ProposalNavigation", () => {
...
-      const po = renderComponent({
-        currentProposalId: 1n,
-        proposalIds: [0n, 1n, 2n],
-        selectProposal: selectProposalSpy,
-      });
+    await po.clickNext();
 
-      await po.clickPrevious();
+    expect(selectProposalSpy).toHaveBeenCalledTimes(1);
+    expect(selectProposalSpy).toHaveBeenCalledWith(0n);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts:95]: The next ID after `1n`, is `2n`, right? Why do we expect `0n`?

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts [hunk 5]
```diff
@@ -81,33 +81,31 @@ describe("ProposalNavigation", () => {
...
-
-      await po.clickNext();
-
-      expect(selectProposalSpy).toHaveBeenCalledTimes(1);
-      expect(selectProposalSpy).toHaveBeenCalledWith(2n);
+  it("should emmit next click", async () => {
+    const selectProposalSpy = jest.fn();
+    const po = renderComponent({
+      currentProposalId: 1n,
+      proposalIds: [0n, 1n, 2n],
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts:88]: Would it make sense to use non-consecutive IDs in the test to catch if we accidentally just return `currentId + 1` instead of the actual next ID? etc.

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts [hunk 6]
```diff
@@ -81,33 +81,31 @@ describe("ProposalNavigation", () => {
...
+      selectProposal: selectProposalSpy,
     });
+
+    await po.clickPrevious();
+
+    expect(selectProposalSpy).toHaveBeenCalledTimes(1);
+    expect(selectProposalSpy).toHaveBeenCalledWith(2n);
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts:109]: same

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts (hunk 5 above)
Comment (dskloetd) [reply frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts:88]: Also, please add more IDs. If there's only a single ID smaller than `1n`, the test doesn't distinguish between getting the smallest or largest ID smaller than `1n`.

Comment (mstrasinskis) [reply frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts]: There is now a separate test for non-consecutive IDs.

**Review by dskloetd — APPROVED** [review]
Thanks!

File: frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts (hunk 2 above)
Comment (dskloetd) [reply frontend/src/tests/lib/components/proposal-detail/ProposalNavigation.spec.ts]: For the record. The problem was that the IDs were passed in increasing order while they needed to be passed in in decreasing order.
